samza-core/src/main/java/org/apache/samza/container/host/OshiBasedStatisticsGetter.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.container.host;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.NotImplementedException;
+import oshi.SystemInfo;
+import oshi.software.os.OSProcess;
+import oshi.software.os.OperatingSystem;
+
+
+/**
+ * An implementation of {@link SystemStatisticsGetter} that relies on using oshi framework(https://www.oshi.ooo/)
+ */
+public class OshiBasedStatisticsGetter implements SystemStatisticsGetter {
+  // the snapshots of current JVM process and its child processes
+  private final Map<Integer, OSProcess> previousProcessSnapshots = new HashMap<>();
+
+  private final OperatingSystem os;
+  private final int cpuCount;
+
+  public OshiBasedStatisticsGetter() {
+    this(new SystemInfo());
+  }
+
+  @VisibleForTesting
+  OshiBasedStatisticsGetter(SystemInfo si) {
+    this(si.getOperatingSystem(), si.getHardware().getProcessor().getPhysicalProcessorCount());
+  }
+
+  @VisibleForTesting
+  OshiBasedStatisticsGetter(OperatingSystem os, int cpuCount) {
+    this.os = os;
+    this.cpuCount = cpuCount;
+  }
+
+  @Override
+  public SystemMemoryStatistics getSystemMemoryStatistics() {
+    throw new NotImplementedException("Not implemented");
+  }
+
+  @Override
+  public ProcessCPUStatistics getProcessCPUStatistics() {
+    final List<OSProcess> currentProcessAndChildProcesses = getCurrentProcessAndChildProcesses();
+    final double totalCPUUsage = getTotalCPUUsage(currentProcessAndChildProcesses);
+    refreshProcessSnapshots(currentProcessAndChildProcesses);
+    return new ProcessCPUStatistics(100d * totalCPUUsage / cpuCount);
```

Comment:
The existing CPU reporting actually takes into account the core-count on the host, right? Because it leverages on Java function `OperatingSystemMXBean. getProcessCpuLoad ()` whose value is a double in the [0.0,1.0] interval. If not, the value could be larger than 1.0.

https://docs.oracle.com/javase/8/docs/jre/api/management/extension/com/sun/management/OperatingSystemMXBean.html#getProcessCpuTime--

>Returns the "recent cpu usage" for the Java Virtual Machine process. This value is a double in the [0.0,1.0] interval. A value of 0.0 means that none of the CPUs were running threads from the JVM process during the recent period of time observed, while a value of 1.0 means that all CPUs were actively running threads from the JVM 100% of the time during the recent period being observed. Threads from the JVM include the application threads as well as the JVM internal threads. All values betweens 0.0 and 1.0 are possible depending of the activities going on in the JVM process and the whole system. If the Java Virtual Machine recent CPU usage is not available, the method returns a negative value.

Reply:
from the desc above, it seems the OperatingSystemMXBean is actually giving out the **fraction** of the total available cpu that the jvm process is using. hence it is between [0,1]. My understanding is based on the following excerpt pasted above

>A value of 0.0 means that none of the CPUs were running threads from the JVM process during the recent period of time observed, while a value of 1.0 means that all CPUs were actively running threads from the JVM 100% of the time during the recent period being observed.

this to me indicates fraction of the num CPUs being used by jvm process. 
Also matches the way `getProccessCpuLoad` is used within Samza's JVMMetrics

```
      val processCpuLoad = operatingSystemMXBean.getProcessCpuLoad
      gProcessCpuUsage.set(processCpuLoad * PCT)
      gProcessCpuUsageProcessors.set(processCpuLoad * operatingSystemMXBean.getAvailableProcessors)
      gSystemCpuUsage.set(operatingSystemMXBean.getSystemCpuLoad * PCT)
```

Note how we multiply processCpuLoad with 100 to get the percent of total cpu being used by the process.

Reply:
Yeah, your understanding is the same as mine. So you are good with this PR's implementation, right?

Reply:
no, i am saying that the current samza process-cpu usage does NOT take into account the num cpu cores on the host. 

OperatingSystemMXBean is actually giving out the fraction of the total available cpu that the jvm process is using. hence it is between [0,1] AND NOT the actual cpu tick count which would need to take into account cpu core count

but the proposed change in the pr takes into account num cores and gives the actual cpu tick count

Reply:
Oh, I see the gap here: Your point is that host CPU num != JVM available CPU num, and I thought they were the same. I think your point is correct. 

I did some research/test on the `OperatingSystemMXBean. getAvailableProcessors()`, the count is for logical CPU count instead of phsical CPU count. For non-containerization or VM use cases, the available logic/phsical CPU count in JVM should be same with host, IIUC. I will update this oshi implementation based on logical CPU count anyway since we should be consistent with JVM. 

Note: I didn't verify if oshi shows the consistent behavior/result regarding the CPU count in the containerizaiton or VM land(e.g. docker, kubernetes).